contracts/wallet.sol

```diff
@@ -661,26 +655,17 @@ contract Wallet is ENSResolvable, Vault, GasTopUpLimit, LoadLimit {
         // This prevents users from accidentally sending Value and Data to a plain old address
         if (_destinationIsContract) {
             require(address(_destination).isContract(), "executeTransaction for a contract: call to non-contract");
-            // Check if there exists at least a method signature in the transaction payload
-            if (_data.length >= 4) {
-                // Get method signature
-                uint32 signature = _data._bytesToUint32(0);
-
-                // Check if method is either ERC20 transfer or approve
-                if (signature == _TRANSFER || signature == _APPROVE) {
-                    require(_data.length >= 4 + 32 + 32, "invalid transfer / approve transaction data");
-                    uint amount = _data._bytesToUint256(4 + 32);
-                    // The "toOrSpender" is the '_to' address for a ERC20 transfer or the '_spender; in ERC20 approve
-                    // + 12 because address 20 bytes and this is padded to 32
-                    address toOrSpender = _data._bytesToAddress(4 + 12);
-
-                    // Check if the toOrSpender is in the whitelist
-                    if (!whitelistMap[toOrSpender]) {
-                        // If the address (of the token contract, e.g) is not in the TokenWhitelist used by
-                        // the convert method, then etherValue will be zero
-                        uint etherValue = convertToEther(_destination, amount);
-                        _spendLimit._enforceLimit(etherValue);
-                    }
+            // Check if the destination contract address is in the tokenWhitelist, hence protected
+            if (_isTokenAvailable(_destination)) {
+                //to is the recipient's address and amount is the value to be transferred
+                address to;
+                uint amount;
+                (to, amount) = _getERC20RecipientAndAmount(_destination, _data);
+                if (!whitelistMap[to]) {
+                    // If the address (of the token contract, e.g) is not in the TokenWhitelist used by the convert method...
+                    // ...then etherValue will be zero
+                    uint etherValue = convertToEther(_destination, amount);
```

Comment:
probably we could save ourselves a bit of gas by returning ETH amount from `_getERC20RecipientAndAmount`

Reply:
Are you suggesting to move convertToEther() to whitelist? not a a bad idea actually...

Reply:
Then we would have one  external call less (_getTokenInfo()) and one extra external call convertToEther() , more or less the same...unless I missed something.